### PR TITLE
feat: retrievals

### PR DESCRIPTION
**Pull request by LexLuthr**

### CODE REVIEW

File: market/retrieval/piecehandler.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	w.Header().Set("Cache-Control", "public, max-age=29030400, immutable")
+}
+
+func serveContent(res http.ResponseWriter, req *http.Request, content io.ReadSeeker) {
+	// Note that the last modified time is a constant value because the data
+	// in a piece identified by a cid will never change.
+
+	if req.Method == http.MethodHead {
+		// For an HTTP HEAD request ServeContent doesn't send any data (just headers)
+		http.ServeContent(res, req, "", time.Time{}, nil)
```
Comment (magik6k) [inline market/retrieval/piecehandler.go:93]: Won't this send incorrect content-length? Or does it not send that header if data is nil?

File: market/retrieval/remoteblockstore/metric.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+import (
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+
+	"github.com/filecoin-project/lotus/metrics"
+)
+
+// Distribution
+var defaultMillisecondsDistribution = view.Distribution(0.01, 0.05, 0.1, 0.3, 0.6, 0.8, 1, 2, 3, 4, 5, 6, 8, 10, 13, 16, 20, 25, 30, 40, 50, 65, 80, 100, 130, 160, 200, 250, 300, 400, 500, 650, 800, 1000, 2000, 3000, 4000, 5000, 7500, 10000, 20000, 50000, 100000)
```
Comment (magik6k) [inline market/retrieval/remoteblockstore/metric.go:11]: This is way too much imo, prometheus doesn't like that much arity in data, would maybe do 30% of those

File: market/retrieval/remoteblockstore/reader.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		}
+		unsealed, err := ro.pp.IsUnsealed(ctx, sr, storiface.UnpaddedByteIndex(dl.Offset), dl.Length)
+		if err != nil {
+			merr = multierror.Append(merr, err)
+			continue
+		}
+
+		if !unsealed {
+			continue
+		}
```
Comment (magik6k) [inline market/retrieval/remoteblockstore/reader.go:105]:
```suggestion
```

This is redundant and will only slow things down, `pp.ReadPiece->tryReadUnsealedPiece->storage.Reader` already does this check internally

File: market/retrieval/remoteblockstore/remoteblockstore.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+var log = logging.Logger("remote-blockstore")
+
+type idxAPI interface {
+	PiecesContainingMultihash(ctx context.Context, m multihash.Multihash) ([]cid.Cid, error)
+	GetOffsetSize(ctx context.Context, pieceCid cid.Cid, hash multihash.Multihash) (*indexstore.OffsetSize, error)
+}
+
+type pieceProviderAPI interface {
+	IsUnsealed(ctx context.Context, sector storiface.SectorRef, offset storiface.UnpaddedByteIndex, size abi.UnpaddedPieceSize) (bool, error)
```
Comment (magik6k) [inline market/retrieval/remoteblockstore/remoteblockstore.go:41]: Ideally we don't need this one

File: market/retrieval/remoteblockstore/remoteblockstore.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+			}(reader)
+
+			// Get the offset of the block within the piece (CAR file)
+			offsetSize, err := ro.idxApi.GetOffsetSize(ctx, pieceCid, c.Hash())
+			if err != nil {
+				return nil, fmt.Errorf("getting offset/size for cid %s in piece %s: %w", c, pieceCid, err)
+			}
+
+			// Seek to the section offset
+			readerAt := readerutil.NewReadSeekerFromReaderAt(reader, int64(offsetSize.Offset))
```
Comment (magik6k) [inline market/retrieval/remoteblockstore/remoteblockstore.go:166]: Pretty sure this thing is like `io.NewSectionReader()` except duplicate

File: market/retrieval/remoteblockstore/remoteblockstore.go [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+			// Read the block data
+			bufferSize := 4096
```
Comment (magik6k) [inline market/retrieval/remoteblockstore/remoteblockstore.go:168]: Why do we not just read the correct size here?

**Review by magik6k — COMMENTED** [review]
Looks pretty good, couple comments to address

File: documentation/en/configuration/default-curio-configuration.md [hunk 7]
```diff
@@ -734,7 +740,13 @@ description: The default curio configuration
 
 
 [HTTP]
-  # DomainName specifies the domain name that the server uses to serve HTTP requests.
+  # Enable the HTTP server on the node
+  #
+  # type: bool
+  #Enable = false
```
[the remark this answers is not in the document]
Comment (LexLuthr) [reply documentation/en/configuration/default-curio-configuration.md:746]: Removed the market duplicate. We will be using this HTTP server for everything so putting it under market seems unfair. Default to True might not be a good in this case. We don't want to start HTTP server on all nodes. Just select nodes user chooses.

File: market/retrieval/piecehandler.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		w.WriteHeader(http.StatusBadRequest)
+		stats.Record(ctx, remoteblockstore.HttpPieceByCid400ResponseCount.M(1))
+		return
+	}
+
+	// Get a reader over the piece
+	reader, err := rp.bs.GetSharedPieceReader(ctx, pieceCid)
+	if err != nil {
+		log.Errorf("server error getting content for piece CID %s: %s", pieceCid, err)
+		if strings.Contains(strings.ToLower(err.Error()), "no deals found") {
```
[the remark this answers is not in the document]
Comment (LexLuthr) [reply market/retrieval/piecehandler.go:48]: I will convert it into a proper error.

File: market/retrieval/piecehandler.go (hunk 1 above)
Comment (LexLuthr) [reply market/retrieval/piecehandler.go:93]: I have set the content length later. Head request should not send Content-Length header.

File: market/retrieval/remoteblockstore/reader.go (hunk 3 above)
Comment (LexLuthr) [reply market/retrieval/remoteblockstore/reader.go:105]: Removed this.

File: market/retrieval/remoteblockstore/remoteblockstore.go (hunk 4 above)
Comment (LexLuthr) [reply market/retrieval/remoteblockstore/remoteblockstore.go:41]: removed

File: market/retrieval/remoteblockstore/remoteblockstore.go (hunk 6 above)
Comment (LexLuthr) [reply market/retrieval/remoteblockstore/remoteblockstore.go:168]: That was a work around for 2k sectors in Boost. Now that we can reproduce the pieceReader panic, I have removed the custom buffer size and went with default.

File: market/retrieval/remoteblockstore/metric.go (hunk 2 above)
Comment (LexLuthr) [reply market/retrieval/remoteblockstore/metric.go:11]: Dumped most of these.

**Review by magik6k — APPROVED** [review]

**Pull request merged**